src/LBHooksBaseSimpleRewarder.sol

```diff
@@ -0,0 +1,146 @@
+    }
+
+    /**
+     * @dev Returns the remaining rewards
+     * @return remainingRewards The remaining rewards
+     */
+    function getRemainingRewards() external view virtual override returns (uint256 remainingRewards) {
+        uint256 balance = TokenHelper.safeBalanceOf(_getRewardToken(), address(this));
+        return balance - _totalUnclaimedRewards - _getPendingTotalRewards();
+    }
+
+    /**
+     * @dev Sets the rewarder parameters
+     * @param maxRewardPerSecond The maximum reward per second
+     * @param startTimestamp The start timestamp
+     * @param expectedDuration The expected duration
+     * @return rewardPerSecond The reward per second
+     */
+    function setRewarderParameters(uint256 maxRewardPerSecond, uint256 startTimestamp, uint256 expectedDuration)
+        external
+        virtual
+        override
+        onlyOwner
+        returns (uint256 rewardPerSecond)
+    {
+        return _setRewardParameters(maxRewardPerSecond, startTimestamp, expectedDuration);
+    }
+
+    /**
+     * @dev Sets the reward per second
+     * @param maxRewardPerSecond The maximum reward per second
+     * @param expectedDuration The expected duration
+     * @return rewardPerSecond The reward per second
+     */
+    function setRewardPerSecond(uint256 maxRewardPerSecond, uint256 expectedDuration)
+        external
+        virtual
+        override
+        onlyOwner
+        returns (uint256 rewardPerSecond)
+    {
+        uint256 lastUpdateTimestamp = _lastUpdateTimestamp;
+        uint256 startTimestamp = lastUpdateTimestamp > block.timestamp ? lastUpdateTimestamp : block.timestamp;
+
+        return _setRewardParameters(maxRewardPerSecond, startTimestamp, expectedDuration);
+    }
+
+    /**
+     * @dev Internal function to set the rewarder parameters
+     * @param maxRewardPerSecond The maximum reward per second
+     * @param startTimestamp The start timestamp
+     * @param expectedDuration The expected duration
+     * @return rewardPerSecond The reward per second
+     */
+    function _setRewardParameters(uint256 maxRewardPerSecond, uint256 startTimestamp, uint256 expectedDuration)
+        internal
+        virtual
+        returns (uint256 rewardPerSecond)
+    {
+        if (startTimestamp < block.timestamp) revert LBHooksBaseSimpleRewarder__InvalidStartTimestamp();
+        if (!_isLinked()) revert LBHooksBaseSimpleRewarder__Stopped();
+        if ((expectedDuration == 0) != (maxRewardPerSecond == 0)) revert LBHooksBaseSimpleRewarder__InvalidDuration();
+
+        _updateAccruedRewardsPerShare();
+
+        uint256 remainingReward = TokenHelper.safeBalanceOf(_getRewardToken(), address(this)) - _totalUnclaimedRewards;
+        uint256 maxExpectedReward = maxRewardPerSecond * expectedDuration;
+
+        rewardPerSecond = maxExpectedReward > remainingReward ? remainingReward / expectedDuration : maxRewardPerSecond;
+        uint256 expectedReward = rewardPerSecond * expectedDuration;
```

Comment:
What's the reason behind updating `rewardPerSecond` according to the remaining rewards vs reverting if there is no sufficient rewards left ? For ops reasons ? Would probably be worth explaining this functionnality in the `setRewarderParameters` and `setRewardPerSecond` natspecs.

Reply:
The main reason is to make it easier to use `setRewardPerSecond` as the time you'd use to calculate the rewards and the time where the transaction is executed will most likely be different; and if the reward per sec is currently not 0, it would revert unless you put a lower amount but you won't be able to distribute all the rewards.
So the idea is to recalculate the rewards to avoid headache when wanting to update the reward per second.
Are you suggesting to add a `minRewardPerSecond` to avoid it becoming too tiny on admin error? (like if you use the end timestamp and not a duration)

Reply:
No no I think that's fine, just wanted to point out that it should be explained in the natspecs